pandas/core/arrays/numpy_.py

```diff
@@ -236,7 +237,13 @@ def __setitem__(self, key, value):
             value = np.asarray(value)
 
         values = self._ndarray
-        t = np.result_type(value, values)
+        if isinstance(value, str):
+            if is_object_dtype(self.dtype._dtype):
+                t = np.dtype(object)
+            else:
+                t = self.dtype._dtype
```

Comment:
`test_setitem_object_typecode[None]` hits it (setting a string into an integer array).

Reply:
simpler to leave the original code then just convert a np.str to no.object (which is what we do inside blocks manager and other places); maybe have a function to do this rather than rewriting logic all over the place

Reply:
I don't think that's appropriate for PandasArray. The idea is to take an arbitrary numpy array and box it in an extension array.

Reply:
and that’s exactly what is done in ObjectBlock now

pls refactor rather than adding logic

Reply:
Can you clarify your proposal then? IIUC, you're saying take

```
PandasArray(np.array(['a', 'b'], dtype='str'))
```

and translate it to the equivalent of

```
PandasArray(np.array(['a', 'b'], dtype="object"))
```

If so, then I'm against doing that.

Reply:
@jreback is the objection here similar to #28150?

Reply:
umm, no, though that issue is interesting as well (we don't have a well defined policy if we are handed a view)

Reply:
looking at this again I am puzzled why you have this

```
if is_object_dtype(self.dtype._dtype):
                t = np.dtype(object)
            else:
                t = self.dtype._dtype
```

why is this not just always ``object``? e.g. what hits the other path?

Reply:
Answered in https://github.com/pandas-dev/pandas/pull/28119/files#r319652823

> `test_setitem_object_typecode[None]` hits it (setting a string into an integer array).

Reply:
Let me try re-summarizing.

The purpose of this block is to determine if we need to cast the backing array to a new dtype. When `self._ndarray.dtype` doesn't match the dtype of the new value, we cast the array to the dtype new value.

The typically works, except for certain strings, which NumPy interprets as typecods.

Do we want to just not do that casting? Do we want to try coercing the `value` to the type of the backing ndarray?

Reply:
> need to cast the backing array to a new dtype

I maintain we should not be doing this (#28150).  Do any tests or anything user-facing rely on this?

Reply:
I doubt it. I think I was mimicking the behavior of `Series.__setitem__`

```python
In [4]: x = np.array([1, 2, 3])

In [5]: s = pd.Series(x)

In [6]: s.values is x
Out[6]: True

In [7]: s[0] = 'a'

In [8]: s.values is x
Out[8]: False
```

But I'm happy to be stricter here.

Reply:
That said, we'll also inherit things like

```python
In [11]: x = np.array([1, 2, 3])

In [12]: x[0] = 5.5

In [13]: x
Out[13]: array([5, 2, 3])
```

But maybe that's OK, if the intent is to be close to NumPy here.

Reply:
To what extent can we punt on the float treatment for now?  I think there's a case to be made that we should raise instead of casting there, but don't want to bog this down any more.

Reply:
Hmm I think our options are to always raises, or adopt NumPy's behavior. I don't think I have a preference.

Reply:
The thought that pushes me towards raising is that if/when this is backing a Block, we want `Block.setitem` to try to set it on block.values and then fall back to casting.